### PR TITLE
Implement For .. of 

### PR DESCRIPTION
**Pull request by mohanson**
```
for (let v of <array>) {

}
```

See Tests:

```ts
let globalarr = [1, 2, 3];

function main(): number {
    let localearr = [4, 5, 6];
    let s: number = 0;
    for (let v of globalarr) {
        s += v;
    }
    for (let v of localearr) {
        s += v;
    }
    return s; // 21
}
```

### CODE REVIEW

File: src/codegen/array-literal-expression.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+export default class CodeGenArray {
+  private cgen: LLVMCodeGen;
+
+  constructor(cgen: LLVMCodeGen) {
+    this.cgen = cgen;
+  }
+
+  public genArrayType(node: ts.ArrayLiteralExpression): llvm.ArrayType {
+    const length = node.elements.length;
+    const elementType = (() => {
```
Comment (yejiayu) [inline src/codegen/array-literal-expression.ts]: length === 0 ? genType  : genExpr

File: src/codegen/array-literal-expression.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        );
+      } else {
+        return this.cgen.genExpression(node.elements[0]).type;
+      }
+    })();
+    return llvm.ArrayType.get(elementType, length);
+  }
+
+  public genArrayInitializer(node: ts.ArrayLiteralExpression): llvm.Value[] {
+    return node.elements.map(item => {
```
Comment (yejiayu) [inline src/codegen/array-literal-expression.ts]:
```suggestion
    return node.elements.map(item => this.cgen.genExpression(item));
```

File: src/codegen/array-literal-expression.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  public genArrayLiteral(node: ts.ArrayLiteralExpression): llvm.AllocaInst {
+    const arrayType = this.genArrayType(node);
+    const arraySize = llvm.ConstantInt.get(
+      this.cgen.context,
+      arrayType.numElements,
+      64
+    );
+    const arrayPtr = this.cgen.builder.createAlloca(arrayType, arraySize);
+    this.genArrayInitializer(node).forEach((item, i) => {
+      const ptr = this.cgen.builder.createInBoundsGEP(arrayPtr, [
```
Comment (yejiayu) [inline src/codegen/array-literal-expression.ts]: Does that mean to assign `0` to every element with index `i`?

File: src/codegen/for-of-statement.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import llvm from 'llvm-node';
+import ts from 'typescript';
+
+import LLVMCodeGen from '.';
```
Comment (yejiayu) [inline src/codegen/for-of-statement.ts]: Usually, we use `./` here.

File: src/codegen/for-of-statement.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+export default class CodeGenForOf {
+  private cgen: LLVMCodeGen;
+
+  constructor(cgen: LLVMCodeGen) {
+    this.cgen = cgen;
+  }
+
+  public genForOfStatement(node: ts.ForOfStatement): void {
+    const i = (() => {
```
Comment (yejiayu) [inline src/codegen/for-of-statement.ts]: Why use an anonymous function?

File: src/codegen/for-statement.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+export default class CodeGenFor {
+  private cgen: LLVMCodeGen;
+
+  constructor(cgen: LLVMCodeGen) {
+    this.cgen = cgen;
+  }
+
+  public genForStatement(node: ts.ForStatement): void {
+    if (node.initializer) {
```
Comment (yejiayu) [inline src/codegen/for-statement.ts]:
```suggestion
    if (node.initializer && ts.isVariableDeclarationList(node.initializer))
```

Avoid nesting.

Comment (mohanson) [reply src/codegen/for-statement.ts]: node.initializer could be an expression. Although I have not added this branch judgment, it should be added later might.

File: src/codegen/for-of-statement.ts (hunk 5 above)
Comment (mohanson) [reply src/codegen/for-of-statement.ts]: both ok.

File: src/codegen/for-of-statement.ts (hunk 4 above)
Comment (mohanson) [reply src/codegen/for-of-statement.ts]: Fixed.

File: src/codegen/array-literal-expression.ts (hunk 3 above)
Comment (mohanson) [reply src/codegen/array-literal-expression.ts]: It's a bug here, it will paniced at empty array, fixed.

Comment (mohanson) [reply src/codegen/array-literal-expression.ts]: Yes. An empty sized array has no meanings.

File: src/codegen/array-literal-expression.ts (hunk 2 above)
Comment (mohanson) [reply src/codegen/array-literal-expression.ts]: Fixed

File: src/codegen/array-literal-expression.ts (hunk 1 above)
Comment (mohanson) [reply src/codegen/array-literal-expression.ts]: both ok

**Pull request merged**